### PR TITLE
Do not store witness txn in rejection cache

### PR DESCRIPTION
**Pull request by sipa**
As proposed in the IRC meeting from august 4th (see http://www.erisian.com.au/meetbot/bitcoin-core-dev/2016/bitcoin-core-dev.2016-08-04-19.00.log.html), this changes the logic to never store witness transactions (or transactions whose witness data was stripped) in the rejection cache.

As an optimization, only double check for witness invalidity if the transaction has no witness. This changes the meaning for state.CorruptionPossible() for transactions from "witness possibly invalidated" to "witness stripped". This is faster (no need for 3 script checks) and simpler for the DoS logic.

This also removes the special logic to not assign DoS for witness failures for non-witness peers. That logic already exists in more generic form, through the non-mandatory script verify flags.

See also #8279.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
@sipa I added a milestone. Hope this is correct.

File: src/main.cpp [hunk 1]
```diff
@@ -5488,7 +5488,10 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                         // Probably non-standard or insufficient fee/priority
                         LogPrint("mempool", "   removed orphan tx %s\n", orphanHash.ToString());
                         vEraseQueue.push_back(orphanHash);
-                        if (!stateDummy.CorruptionPossible()) {
+                        if (orphanTx.wit.IsNull() && !state.CorruptionPossible()) {
```
Comment (instagibbs) [inline src/main.cpp]: s/state/stateDummy/ ?

Comment (sipa) [inline src/main.cpp]: Nice catch, fixed!

**Comment by instagibbs** [comment]
code review ACK https://github.com/bitcoin/bitcoin/pull/8525/commits/59d0df0193181a95f76f0d1c13f311461457aa03

**Comment by laanwj** [comment]
concept ACK https://github.com/bitcoin/bitcoin/pull/8525/commits/59d0df0193181a95f76f0d1c13f311461457aa03

**Comment by sipa** [comment]
@sdaftuar @morcos Comments?

**Comment by instagibbs** [comment]
created a tiny test for the witness-stuffing case that previous blinded nodes: https://github.com/instagibbs/bitcoin/commit/62778b20a971653cd84d2797adf51720985f55a6

**Comment by sipa** [comment]
Rebased, and incorporated @instagibbs' test.

**Comment by laanwj** [comment]
utACK ca10a03

**Comment by sdaftuar** [comment]
@sipa Sorry for the late review!  One question about this:

> As an optimization, only double check for witness invalidity if the transaction has no witness. This changes the meaning for state.CorruptionPossible() for transactions from "witness possibly invalidated" to "witness stripped". This is faster (no need for 3 script checks) and simpler for the DoS logic.

We still have one place in AcceptToMemoryPoolWorker where we return a state with CorruptionPossible set to true, even when there is a witness:

https://github.com/sipa/bitcoin/blob/ca10a03addf70421893791c2c499e82fc494d60b/src/main.cpp#L1147

This doesn't seem to be a problem at all, because the caller is checking to see if the witness is NULL before checking CorruptionPossible, but I was wondering if the intention was to eliminate this type of usage?

**Comment by luke-jr** [comment]
Do we need this in 0.13?

**Comment by luke-jr** [comment]
Sorry, missed the milestone somehow

**Pull request merged**